app/main.py

```
@@ -1,44 +1,58 @@
+import asyncio
 import time
+from typing import Any, Awaitable
 
-from iot.devices import HueLightDevice, SmartSpeakerDevice, SmartToiletDevice
-from iot.message import Message, MessageType
-from iot.service import IOTService
+from app.iot.devices import HueLightDevice, SmartSpeakerDevice, SmartToiletDevice
+from app.iot.message import Message, MessageType
+from app.iot.service import IOTService
 
 
-def main() -> None:
+async def run_sequence(*functions: Awaitable[Any]) -> None:
+    for function in functions:
+        await function
+
+
+async def run_parallel(*functions: Awaitable[Any]) -> None:
+    await asyncio.gather(*functions)
+
+
+async def main() -> None:
     # create an IOT service
     service = IOTService()
 
     # create and register a few devices
     hue_light = HueLightDevice()
     speaker = SmartSpeakerDevice()
     toilet = SmartToiletDevice()
-    hue_light_id = service.register_device(hue_light)
-    speaker_id = service.register_device(speaker)
-    toilet_id = service.register_device(toilet)
-
-    # create a few programs
-    wake_up_program = [
-        Message(hue_light_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.SWITCH_ON),
-        Message(speaker_id, MessageType.PLAY_SONG, "Rick Astley - Never Gonna Give You Up"),
-    ]
-
-    sleep_program = [
-        Message(hue_light_id, MessageType.SWITCH_OFF),
-        Message(speaker_id, MessageType.SWITCH_OFF),
-        Message(toilet_id, MessageType.FLUSH),
-        Message(toilet_id, MessageType.CLEAN),
-    ]
 
-    # run the programs
-    service.run_program(wake_up_program)
-    service.run_program(sleep_program)
+    hue_light_id, speaker_id, toilet_id = await asyncio.gather(
+        service.register_device(hue_light),
+        service.register_device(speaker),
+        service.register_device(toilet),
+    )
```

Suggestion: device registration is done concurrently using `asyncio.gather` here which is correct. To improve robustness consider handling partial failures (for example `await asyncio.gather(..., return_exceptions=True)` and then inspect results, or wrap the gather in try/except) so a failing register doesn't hide other errors. This is optional but improves error visibility during concurrent registration.